backend/tools/keycloak/sync/Factory.cs

```diff
@@ -0,0 +1,386 @@
+        /// Creates a new instance of an Factory class, initializes it with the specified arguments.
+        /// </summary>
+        /// <param name="options"></param>
+        /// <param name="auth"></param>
+        /// <param name="clientFactory"></param>
+        /// <param name="tokenHandler"></param>
+        /// <param name="logger"></param>
+        public Factory(IOptionsMonitor<ToolOptions> options, IOpenIdConnector auth, IHttpClientFactory clientFactory, JwtSecurityTokenHandler tokenHandler, ILogger<Factory> logger)
+        {
+            _options = options.CurrentValue;
+            _auth = auth;
+            _client = clientFactory.CreateClient("Pims.Tools.Keycloak.Sync");
+            _tokenHandler = tokenHandler;
+            _logger = logger;
+        }
+        #endregion
+
+        #region Methods
+        /// <summary>
+        /// Sync roles, groups and users in keycloak and PIMS.
+        /// </summary>
+        /// <returns></returns>
+        public async Task<int> SyncAsync()
+        {
+            // Check if token has expired.  If it has refresh it.
+            if (String.IsNullOrWhiteSpace(_token) || _tokenHandler.ReadJwtToken(_token).ValidTo <= DateTime.UtcNow)
+            {
+                var tokenNew = await _auth.RequestTokenAsync(_refreshToken);
+                _token = tokenNew.access_token;
+                _refreshToken = tokenNew.refresh_token;
+            }
+
+            // Fetch all Claims in PIMS and link or create them in keycloak.
+            var claims = await HandleRequestAsync<PageModel<ClaimModel>>(HttpMethod.Get, $"{_options.Api.Uri}/admin/claims?page=1&quantity=50"); // TOD: Handle paging.
+            foreach (var claim in claims?.Items)
+            {
+                var kclaim = await GetClaimAsync(claim);
+                claim.KeycloakRoleId = kclaim.id;
+
+                _logger.LogInformation($"Updating Claim '{claim.Name}'.");
+                await HandleRequestAsync<ClaimModel, ClaimModel>(HttpMethod.Put, $"{_options.Api.Uri}/admin/claims/{claim.Id}", claim);
+            }
+
+            // Fetch all Roles in PIMS and link or create them in keycloak.
+            var roles = await HandleRequestAsync<PageModel<RoleModel>>(HttpMethod.Get, $"{_options.Api.Uri}/admin/roles?page=1&quantity=50"); // TODO: Handle paging.
+            foreach (var role in roles?.Items)
+            {
+                var krole = await GetRoleAsync(role);
+                role.KeycloakGroupId = krole.id;
+
+                _logger.LogInformation($"Updating Role '{role.Name}'.");
+                await HandleRequestAsync<RoleModel, RoleModel>(HttpMethod.Put, $"{_options.Api.Uri}/admin/roles/{role.Id}", role);
+            }
+
+            // Fetch all Users in PIMS and update keycloak so they have the same roles and claims.
+            var users = await HandleRequestAsync<PageModel<UserModel>>(HttpMethod.Get, $"{_options.Api.Uri}/admin/users?page=1&quantity=50"); // TODO: Handle paging.
```

Comment:
So this is fetching the top 50 users in PIMS and updating their roles and claims? I suppose that works for now.

Reply:
Yes for a once off critical job for today, it works.